File: code_assistent_Agent/server/utils/agent_result.ts
export type AgentSource = {
  type: 'url' | 'doc' | 'table' | 'sql'
  ref: string
}

/** 与 Manager N2 / U2 对齐的标准失败码 */
export type CodeErrorCode =
  | 'timeout'
  | 'empty_result'
  | 'needs_clarify'
  | 'business'
  | 'tool_round_limit'
  | 'http_5xx'

export type AgentResult = {
  ok: boolean
  agent: string
  trace_id?: string
  answer?: string
  sources?: AgentSource[]
  structured?: Record<string, unknown>
  error_code?: string
  latency_ms?: number
  /** G1：提供商或估算用量，供总管 runBudget */
  usage?: { tokens?: number; usd?: number; actual?: boolean }
}

export function classifyCodeThrownError(error: unknown): CodeErrorCode {
  const msg = String((error as Error)?.message || error || '').toLowerCase()
  if (
    msg.includes('timeout') ||
    msg.includes('timed out') ||
    msg.includes('aborted') ||
    msg.includes('deadline')
  ) {
    return 'timeout'
  }
  if (msg.includes('tool_round_limit') || msg.includes('max tool rounds') || msg.includes('recursion limit')) {
    return 'tool_round_limit'
  }
  if (/\b5\d{2}\b/.test(msg) || msg.includes('http_5xx') || msg.includes('internal server')) {
    return 'http_5xx'
  }
  return 'business'
}

export function buildCodeFailAgentResult(params: {
  error_code: CodeErrorCode | string
  answer?: string
  trace_id?: string
  ms?: number
  structured?: Record<string, unknown>
}): AgentResult {
  return {
    ok: false,
    agent: 'code',
    trace_id: params.trace_id,
    answer: params.answer || '',
    structured: params.structured,
    error_code: String(params.error_code || 'business'),
    latency_ms: params.ms
  }
}

export function buildCodeComputeAgentResult(params: {
  answer: string
  trace_id?: string
  ms?: number
  task_kind?: string
  error_code?: string
  usage?: { tokens?: number; usd?: number; actual?: boolean }
}): AgentResult {
  const answer = String(params.answer || '')
  const ok = Boolean(answer.trim()) && !params.error_code
  return {
    ok,
    agent: 'code',
    trace_id: params.trace_id,
    answer,
    structured: {
      task_kind: params.task_kind || 'compute',
      ms: params.ms
    },
    error_code: ok ? undefined : params.error_code || 'empty_result',
    latency_ms: params.ms,
    ...(params.usage ? { usage: params.usage } : {})
  }
}

export function buildCodeRetrieveAgentResult(params: {
  query: string
  hits: number
  snippets: Array<{ path: string; score?: number }>
  trace_id?: string
  ms?: number
}): AgentResult {
  const sources: AgentSource[] = params.snippets
    .map((s) => String(s.path || '').trim())
    .filter(Boolean)
    .slice(0, 12)
    .map((ref) => ({ type: 'doc' as const, ref }))
  return {
    ok: params.hits > 0,
    agent: 'code',
    trace_id: params.trace_id,
    answer: params.query,
    sources: sources.length ? sources : undefined,
    structured: { hits: params.hits, ms: params.ms },
    error_code: params.hits ? undefined : 'empty_result',
    latency_ms: params.ms
  }
}

File: CodePy_Agent/app/protocol/agent_result.py
"""Manager-compatible agentResult envelope (agent=code)."""

from __future__ import annotations

from typing import Any, TypedDict


class AgentSource(TypedDict):
    type: str
    ref: str


class AgentResult(TypedDict, total=False):
    ok: bool
    agent: str
    trace_id: str
    answer: str
    sources: list[AgentSource]
    structured: dict[str, Any]
    needs_clarify: bool
    clarify_questions: list[str]
    error_code: str
    latency_ms: int
    usage: dict[str, Any]


def classify_code_error(error: Any) -> str:
    msg = str(getattr(error, "message", None) or error or "").lower()
    if any(x in msg for x in ("timeout", "timed out", "aborted", "deadline")):
        return "timeout"
    if any(x in msg for x in ("tool_round_limit", "max tool rounds", "recursion limit")):
        return "tool_round_limit"
    if "http_5xx" in msg or "internal server" in msg:
        return "http_5xx"
    return "business"


def build_code_fail_agent_result(
    *,
    error_code: str,
    answer: str = "",
    trace_id: str | None = None,
    ms: int | None = None,
    structured: dict[str, Any] | None = None,
) -> AgentResult:
    return {
        "ok": False,
        "agent": "code",
        "trace_id": trace_id or "",
        "answer": answer or "",
        "structured": structured or {},
        "error_code": str(error_code or "business"),
        "latency_ms": ms or 0,
    }


def build_code_compute_agent_result(
    *,
    answer: str,
    trace_id: str | None = None,
    ms: int | None = None,
    task_kind: str = "compute",
    error_code: str | None = None,
    usage: dict[str, Any] | None = None,
    extra_structured: dict[str, Any] | None = None,
) -> AgentResult:
    text = str(answer or "")
    ok = bool(text.strip()) and not error_code
    structured: dict[str, Any] = {"task_kind": task_kind or "compute", "ms": ms}
    if extra_structured:
        structured.update(extra_structured)
    result: AgentResult = {
        "ok": ok,
        "agent": "code",
        "answer": text,
        "structured": structured,
        "latency_ms": ms or 0,
    }
    if trace_id:
        result["trace_id"] = trace_id
    if not ok:
        result["error_code"] = error_code or "empty_result"
    if usage:
        result["usage"] = usage
    return result


def build_code_edit_agent_result(
    *,
    answer: str,
    trace_id: str | None = None,
    ms: int | None = None,
    task_kind: str = "edit",
    files_touched: list[str] | None = None,
    validate_ok: bool | None = None,
    unified_diff: str | None = None,
) -> AgentResult:
    text = str(answer or "")
    structured: dict[str, Any] = {
        "task_kind": task_kind,
        "ms": ms,
        "files_touched": files_touched or [],
    }
    if validate_ok is not None:
        structured["validate_ok"] = validate_ok
    if unified_diff:
        structured["unified_diff"] = unified_diff[:8000]
    return {
        "ok": bool(text.strip()),
        "agent": "code",
        "trace_id": trace_id or "",
        "answer": text,
        "structured": structured,
        "latency_ms": ms or 0,
        "sources": [{"type": "doc", "ref": p} for p in (files_touched or [])[:12]],
    }
